### ISSUE TITLE
Investigate if `DistributedContextPropagator` can be used to avoid injecting and extracting context in instrumentation libraries

### ISSUE THREAD

**Issue by vishweshbankwar** [open-telemetry/opentelemetry-dotnet-contrib#2013]
# Feature Request

Starting `.NET6.0`  [DistributedContextPropagator](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.distributedcontextpropagator?view=net-7.0) is added to `System.Diagnostics.DiagnosticSource`. Opening this issue to investigate if we can avoid context injection and extraction in `HttpClient` and `ASP.NET Core` instrumentations respectively and have libraries set the correct context based on the default propagator set by user in SDK. If possible, this will help reduce the overhead of extracting and injecting context twice (first in instrumented library i.e. ASP.NET Core \ HttpClient and then in instrumentation library) when the default propagator is not `TraceContextPropagator`.

**Comment by martincostello** [comment]
I asked Claude to investigate this issue, and it came up with the following.

TL;DR - in .NET 6+ we'd need to make a adapter in the SDK that works for both framework and OTel trace context, assign it as the default propagator, then once the instrumentation depends on a version with that behaviour, we can skip the double extraction.

---

## Investigation: can `DistributedContextPropagator` avoid double inject/extract in HttpClient & ASP.NET Core instrumentation?

### Current design

Both instrumentation libraries use OpenTelemetry's own propagator abstraction (`Propagators.DefaultTextMapPropagator`), **not** the runtime's `System.Diagnostics.DistributedContextPropagator`. Neither library references `DistributedContextPropagator` at all.

**HttpClient — inject** (`HttpHandlerDiagnosticListener.cs`):

```csharp
// Propagate context irrespective of sampling decision
var textMapPropagator = Propagators.DefaultTextMapPropagator;
if (textMapPropagator is not TraceContextPropagator)
{
    textMapPropagator.Inject(new PropagationContext(activity.Context, Baggage.Current), request, ...HeaderValueSetter);
}
```

**ASP.NET Core — extract** (`HttpInListener.cs`): same `is not TraceContextPropagator` guard, then `Extract`, then — if the extracted context differs from the `Activity` the framework already created — it builds a **sibling `Activity`** re-parented to the extracted context, marks the framework `Activity` unsampled, and keeps a reference to it (the gRPC tag-copy workaround for #1778).

### Why the double work happens

The .NET runtime already propagates context itself using `DistributedContextPropagator.Current` (default = W3C):

- HttpClient's `DiagnosticsHandler` **injects** `traceparent` on every outbound request.
- ASP.NET Core hosting **extracts** the parent and creates the server `Activity` from it.

So when the OTel-configured propagator is anything other than pure W3C, propagation runs **twice**: once by the runtime (W3C), once by the instrumentation (OTel format).

Two nuances worth calling out:

1. **The `is not TraceContextPropagator` fast-path is narrower than it looks.** The SDK's *default* `Propagators.DefaultTextMapPropagator` is a **composite of `TraceContextPropagator` + `BaggagePropagator`**, which is *not* a `TraceContextPropagator`. So out-of-the-box the guard is `true` and the library still injects/extracts — meaning even the default configuration double-injects `traceparent` today (harmless, but real).
2. **The sibling-`Activity` machinery in ASP.NET Core exists only because** the runtime already created the `Activity` via W3C extraction and the library must re-parent it for non-W3C formats.

### Can `DistributedContextPropagator` replace this? Yes, technically — with caveats

**Idea:** install an adapter that subclasses `DistributedContextPropagator`, delegates to `Propagators.DefaultTextMapPropagator`, and assign it to `DistributedContextPropagator.Current`. Then the **runtime performs propagation once, in the OTel-configured format**, and the instrumentation libraries can skip inject/extract entirely.

**Benefits:**

- Eliminates the second inject/extract for non-W3C propagators (the exact overhead this issue targets).
- **Removes the ASP.NET Core sibling-`Activity` creation** and the associated `IsAllDataRequested = false` juggling and #1778 gRPC tag-copy workaround — the runtime would create the server `Activity` with the correct parent directly.
- Fixes the default-config double-`traceparent` injection.

**Caveats / blockers:**

- **API impedance mismatch.** `DistributedContextPropagator` exposes `Inject(Activity, carrier, setter)`, `ExtractTraceIdAndState(...)` (returns *strings*, not a full `ActivityContext`), and a *separate* `ExtractBaggage(...)`. OTel's `TextMapPropagator` works with `PropagationContext` / typed carriers. A shim is writable, but mapping sampling flags (B3 single/multi encode `sampled` differently) and splitting trace vs. baggage extraction needs care.
- **.NET 6+ only.** `DistributedContextPropagator` does not exist on `netfx` / `netstandard2.0`, so those TFMs must keep today's behavior — this is an addition, not a replacement.
- **Process-wide blast radius.** `DistributedContextPropagator.Current` is a global that affects *all* `Activity` propagation in the process, including non-OTel-instrumented libraries. That's arguably desirable, but it is a behavior change that needs opt-in consideration.
- **Right home is the SDK, not contrib.** The natural place to set `DistributedContextPropagator.Current` is `Sdk.SetDefaultTextMapPropagator` in `opentelemetry-dotnet`, so it stays in sync with the OTel propagator. Contrib libraries would then just *stop* injecting/extracting on .NET 6+. This makes it a cross-repo change requiring coordination (and likely a spec discussion, since it diverges from the "instrumentation owns propagation" model).

### Recommendation

Feasible and worthwhile — but the biggest concrete win is **removing the sibling-`Activity` complexity**, not the raw inject/extract cost (which only affects the minority using non-default propagators). Because it means changing a process-global and belongs in the SDK:

1. Prototype a `DistributedContextPropagator`-derived adapter over `TextMapPropagator` (net6.0+).
2. Wire it from `Sdk.SetDefaultTextMapPropagator` in `opentelemetry-dotnet`, behind a flag initially.
3. Gate the contrib libraries to skip their own inject/extract when the adapter is active on .NET 6+, keeping the existing path for `netfx` / `netstandard`.
